connexion/apis/abstract.py

```diff
@@ -238,23 +242,195 @@ def get_request(self, *args, **kwargs):
+
+    @classmethod
+    @abc.abstractmethod
+    def _is_framework_response(cls, response):
+        """ Return True if `response` is a framework response class """
+
+    @classmethod
+    @abc.abstractmethod
+    def _framework_to_connexion_response(cls, response, mimetype):
+        """ Cast framework response class to ConnexionResponse used for schema validation """
+
+    @classmethod
+    @abc.abstractmethod
+    def _connexion_to_framework_response(cls, response, mimetype, extra_context=None):
+        """ Cast ConnexionResponse to framework response class """
+
+    @classmethod
+    @abc.abstractmethod
+    def _build_response(cls, data, mimetype, content_type=None, status_code=None, headers=None, extra_context=None):
         """
-        This method converts the user framework response to a ConnexionResponse.
-        :param response: A response to cast.
+        Create a framework response from the provided arguments.
+        :param data: Body data.
+        :param content_type: The response mimetype.
+        :type content_type: str
+        :param content_type: The response status code.
+        :type status_code: int
+        :param headers: The response status code.
+        :type headers: Union[Iterable[Tuple[str, str]], Dict[str, str]]
+        :param extra_context: dict of extra details, like url, to include in logs
+        :type extra_context: Union[None, dict]
+        :return A framework response.
+        :rtype Response
         """
 
+    @classmethod
+    def _prepare_body_and_status_code(cls, data, mimetype, status_code=None, extra_context=None):
+        if data is NoContent:
+            data = None
+
+        if status_code is None:
+            if data is None:
+                status_code = 204
+                mimetype = None
+            else:
+                status_code = 200
+        elif hasattr(status_code, "value"):
+            # If we got an enum instead of an int, extract the value.
+            status_code = status_code.value
+
+        if data is not None:
+            body = cls._jsonify_data(data, mimetype)
+        else:
+            body = data
+
+        if extra_context is None:
+            extra_context = {}
+        logger.debug('Prepared body and status code (%d)',
+                     status_code,
+                     extra={
+                         'body': body,
+                         **extra_context
+                     })
+
+        return body, status_code
+
+    @classmethod
+    def _jsonify_data(cls, data, mimetype):
+        if not isinstance(data, bytes):
+            # assume mimetype is json if mimetype is None and it looks like json
+            if (mimetype is None and isinstance(data, (dict, list, tuple))) \
+                    or (isinstance(mimetype, str) and is_json_mimetype(mimetype)):
+                body = cls.jsonifier.dumps(data)
+            elif isinstance(data, str):
+                body = data
+            else:
+                body = str(data)
```

Comment:
What are some other datatypes where the string representation is better? I'm having trouble thinking of any.
For both `int` and `float`, I think I'd prefer the json representation.

Reply:
ok. Now, when mimetype is None, anything that can be jsonified will be, catching TypeError to rescue with `str()`. Duck-typing for the win - if it quacks like JSON and the mimetype is None, then it is JSON.

Reply:
I think this is getting really close.
Eventually we want to be able to register serializers for arbitrary mimetypes, so I think there's a default path (mimetype is none, try to do json), and then maybe some built-in overridable serializers (text/plain uses str() for example).
So if we can get backwards compatible behavior with flask (default serializer is json), but an escape hatch registering new mimetypes/serializers then I think we're golden.
One thing I don't think we should do is introduce anything opinionated, or anything too magical/surprising.

Reply:
It turns out that it is magical and surprising no matter how I look at it. From both aiohttp and flask sides, it feels a little off. :(